src/theory/bv/rewrites

```diff
@@ -0,0 +1,321 @@
+  (extract high low x))
+(define-cond-rule bv-extract-sign-extend-2
+  ((x ?BitVec) (low Int) (high Int) (k Int))
+  (def (n (@bvsize x)))
+  (and (< low n) (>= high n))
+  (extract high low (sign_extend k x))
+  (sign_extend
+    (+ 1 (- high n))
+    (extract (- n 1) low x)))
+(define-cond-rule bv-extract-sign-extend-3
+  ((x ?BitVec) (low Int) (high Int) (k Int))
+  (def (n (@bvsize x)))
+  (>= low n)
+  (extract high low (sign_extend k x))
+  (repeat (+ 1 (- high low)) (extract (- n 1) (- n 1) x)))
+
+(define-rule bv-neg-mult
+  ((xs ?BitVec) (ys ?BitVec) (n Int) (m Int))
+  (bvneg (bvmul xs (@bv n m) ys))
+  (bvmul xs (@bv (- n) m) ys))
+(define-rule* bv-neg-add
+  ((x ?BitVec) (y ?BitVec) (zs ?BitVec :list))
+  (bvneg (bvadd x y zs))
+  (bvneg (bvadd y zs))
+  (bvadd (bvneg x) _))
+
+(define-rule bv-mult-distrib-const-neg
+  ((x ?BitVec) (n Int) (m Int))
+  (bvmul (bvneg x) (@bv n m))
+  (bvmul x (@bv (- n) m)))
+(define-rule bv-mult-distrib-const-add
+  ((x ?BitVec) (y ?BitVec) (n Int) (m Int))
+  (bvmul (bvadd x y) (@bv n m))
+  (bvadd
+    (bvmul x (@bv n m))
+    (bvmul y (@bv n m))
+  ))
+(define-rule bv-mult-distrib-const-sub
+  ((x ?BitVec) (y ?BitVec) (n Int) (m Int))
+  (bvmul (bvsub x y) (@bv n m))
+  (bvsub
+    (bvmul x (@bv n m))
+    (bvmul y (@bv n m))
+  ))
+
+(define-rule bv-mult-distrib-1
+  ((x1 ?BitVec) (x2 ?BitVec) (y ?BitVec))
+  (bvmul (bvadd x1 x2) y)
+  (bvadd (bvmul x1 y) (bvmul x2 y)))
+(define-rule bv-mult-distrib-2
+  ((x1 ?BitVec) (x2 ?BitVec) (y ?BitVec))
+  (bvmul y (bvadd x1 x2))
+  (bvadd (bvmul y x1) (bvmul y x2)))
+(define-rule bv-not-xor
+  ((x ?BitVec) (xs ?BitVec :list))
+  (bvnot (bvxor x xs))
+  (bvxor (bvnot x) xs))
+
+(define-rule* bv-and-simplify-1
+  ((xs ?BitVec :list) (ys ?BitVec :list) (zs ?BitVec :list) (x ?BitVec))
+  (bvand xs x ys x zs)
+  (bvand xs x ys zs))
+(define-rule* bv-and-simplify-2
+  ((xs ?BitVec :list) (ys ?BitVec :list) (zs ?BitVec :list) (x ?BitVec))
+  (bvand xs x ys (bvnot x) zs)
+  (@bv 0 (@bvsize x)))
+
+(define-rule* bv-or-simplify-1
+  ((xs ?BitVec :list) (ys ?BitVec :list) (zs ?BitVec :list) (x ?BitVec))
+  (bvor xs x ys x zs)
+  (bvor xs x ys zs))
+(define-rule* bv-or-simplify-2
+  ((xs ?BitVec :list) (ys ?BitVec :list) (zs ?BitVec :list) (x ?BitVec))
+  (bvor xs x ys (bvnot x) zs)
+  (bvnot (@bv 0 (@bvsize x))))
+
+(define-rule* bv-xor-simplify-1
+  ((xs ?BitVec :list) (ys ?BitVec :list) (zs ?BitVec :list) (x ?BitVec))
+  (bvxor xs x ys x zs)
+  (bvxor xs ys zs))
+(define-rule* bv-xor-simplify-2
+  ((xs ?BitVec :list) (ys ?BitVec :list) (zs ?BitVec :list) (x ?BitVec))
+  (bvxor xs x ys (bvnot x) zs)
+  (bvnot (bvxor xs ys zs)))
+(define-rule* bv-xor-simplify-3
+  ((xs ?BitVec :list) (ys ?BitVec :list) (zs ?BitVec :list) (x ?BitVec))
+  (bvxor xs (bvnot x) ys x zs)
+  (bvnot (bvxor xs ys zs)))
+
+; -- Simplify Bitblasting --
+
+; Unused
+; -a1 - a2 - ... - an + ak + ..  ==> - (a1 + a2 + ... + an) + ak
+;(define-rule bv-bb-add-neg )
```

Comment:
This should be deleted if there is not reason to keep it here.

Reply:
deleted